polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -0,0 +1,168 @@
+
+#include "mlir/Dialect/Polygeist/Analysis/UniformityAnalysis.h"
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/IntegerRangeAnalysis.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "uniformity-analysis"
+
+using namespace mlir;
+using namespace mlir::dataflow;
+using namespace mlir::polygeist;
+
+//===----------------------------------------------------------------------===//
+// Uniformity
+//===----------------------------------------------------------------------===//
+
+namespace mlir {
+namespace polygeist {
+
+raw_ostream &operator<<(raw_ostream &os, const Uniformity &uniformity) {
+  if (uniformity.isUninitialized())
+    return os << "<UNINITIALIZED>";
+
+  os << "uniformity: ";
+  switch (uniformity.getKind()) {
+  case Uniformity::Kind::Unknown:
+    return os << "unknown\n";
+  case Uniformity::Kind::Uniform:
+    return os << "uniform\n";
+  case Uniformity::Kind::NonUniform:
+    return os << "non-uniform\n";
+  }
+  llvm_unreachable("Unhandled case");
+}
+
+} // namespace polygeist
+} // namespace mlir
+
+void UniformityLattice::onUpdate(DataFlowSolver *solver) const {
+  Lattice::onUpdate(solver);
+
+  auto value = point.get<Value>();
+  Operation *op = value.getDefiningOp();
+
+  if (op) {
+    if (auto branch = dyn_cast<RegionBranchOpInterface>(op)) {
+      for (Value operand : branch->getOperands()) {
+        auto *uniformity = solver->lookupState<UniformityLattice>(operand);
+        assert(uniformity && "Expecting a valid uniformity");
+
+        auto *lattice = solver->getOrCreateState<UniformityLattice>(value);
+        return solver->propagateIfChanged(lattice, lattice->join(*uniformity));
+      }
+    }
+  }
+
+  // Note: required to get the dataflow framework to visit nested regions.
+  auto *cv = solver->getOrCreateState<Lattice<ConstantValue>>(value);
+  return solver->propagateIfChanged(
+      cv, cv->join(ConstantValue::getUnknownConstant()));
+}
+
+void Uniformity::print(raw_ostream &os) const { os << *this; }
+
+//===----------------------------------------------------------------------===//
+// UniformityAnalysis
+//===----------------------------------------------------------------------===//
+
+void UniformityAnalysis::visitOperation(
+    Operation *op, ArrayRef<const UniformityLattice *> operands,
+    ArrayRef<UniformityLattice *> results) {
+  LLVM_DEBUG(llvm::dbgs() << "UA: Visiting operation: " << *op << "\n");
+
+  // If the lattice on any operand is uninitialized, bail out.
+  if (llvm::any_of(operands, [](const UniformityLattice *lattice) {
+        return lattice->getValue().isUninitialized();
+      })) {
+    return;
+  }
+
+  // These operations always yield non-uniform result(s).
+  if (isa<sycl::SYCLItemGetIDOp, sycl::SYCLNDItemGetGlobalIDOp>(op)) {
```

Comment:
Does this cover all relevant cases, including kernels only taking an `id` as argument, i.e.
```
cgh.parallel_for(range<1>{...}, [=](id<1> i){...});
```

Reply:
I think we shouldn't be handling any operation differently here. I propose creating a `UniformityOpInterface` for operations to express whether they define uniform values.

Reply:
> Maybe the correct way to approach this is to assume an operation does not define non-uniform values unless it implements some interface.

Yes we need to assume that operations in other dialects (e.g. an arith constant definition) yield uniform values

> We could then attach this interface to all operations that we know will produce a non-uniform value in the SYCL dialect and other relevant dialects and then basically do taint-tracking, where the "taint" is non-uniform values

Yes, essentially certain operations *always* define non-uniform values even if the input operands (if any) are uniform.  

I am willing to add an interface to our SYCL dialect operations, either in this PR or a subsequent one.

Reply:
> I am willing to add an interface to our SYCL dialect operations, either in this PR or a subsequent one.

Sounds like a good approach, I leave it up to you if you want to add the interface in this PR or in the next part of the analysis.